game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/BattleStep.java

```diff
@@ -0,0 +1,88 @@
+package games.strategy.triplea.delegate.battle.steps;
+
+import games.strategy.engine.delegate.IDelegateBridge;
+import games.strategy.triplea.delegate.ExecutionStack;
+import games.strategy.triplea.delegate.IExecutable;
+import games.strategy.triplea.delegate.battle.BattleActions;
+import games.strategy.triplea.delegate.battle.BattleState;
+import java.util.List;
+import lombok.AllArgsConstructor;
+
+/**
+ * A step in a battle.
+ *
+ * <ol>
+ *   <ul>
+ *     Each step can have 0 or more names. These names are shown in the {@link
+ *     games.strategy.triplea.ui.BattlePanel}
+ *   </ul>
+ *   <ul>
+ *     Each step can also have an executable. See {@link BattleAtomic} on what an executable is
+ *     comprised of.
+ *   </ul>
+ * </ol>
+ */
+@AllArgsConstructor
+public abstract class BattleStep {
+
+  /**
+   * The current state of the battle
+   *
+   * <p>Note: This is current mutable so the underlying data can change.
+   */
+  protected final BattleState battleState;
+
+  /** Actions that can occur in a battle that require interaction with {@link IDelegateBridge} */
+  protected final BattleActions battleActions;
+
+  public abstract BattleAtomic getExecutable();
+
+  public abstract List<String> getNames();
+
+  /**
+   * Determine if this step should run based on the request
+   *
+   * @return true if valid
+   */
+  public abstract boolean valid();
+
+  /**
+   * Executes the step
+   *
+   * <p>This is called by the BattleAtomic and {@link #valid} has already been checked
+   *
+   * @param stack The current stack of steps
+   * @param bridge DelegateBridge for interacting with the rest of the program
+   */
+  protected abstract void execute(ExecutionStack stack, IDelegateBridge bridge);
+
+  /**
+   * This is used to break up the battle into separate atomic pieces. If there is a network error,
+   * or some other unfortunate event, then we need to keep track of what pieces we have executed,
+   * and what is left to do. Each atomic step is in its own BattleAtomic with the definition of
+   * atomic is that either:
+   *
+   * <ol>
+   *   <li>The code does not call to an IDisplay, IPlayer, or IRandomSource
+   *   <li>If the code calls to an IDisplay, IPlayer, IRandomSource, and an exception is called from
+   *       one of those methods, the exception will be propagated out of execute() and the execute
+   *       method can be called again.
+   * </ol>
+   */
+  public abstract class BattleAtomic implements IExecutable {
```

Comment:
Is this an inner class only because of `IExecutable`?

Reply:
No, it needs a reference to the BattleStep that is enclosing it so that it can then call `valid` on it and `execute`.

I could change it to have a instance variable that is the BattleStep and each step has to construct it and pass in itself.  The `execute` method on `BattleStep` will need to made public as well.